engines/director/debugger/dt-save-state.cpp

```diff
@@ -0,0 +1,163 @@
+/* ScummVM - Graphic Adventure Engine
+ *
+ * ScummVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "backends/imgui/imgui_utils.h"
+#include "common/formats/json.h"
+#include "common/memstream.h"
+#include "common/savefile.h"
+#include "common/file.h"
+
+#include "director/director.h"
+#include "director/debugger/dt-internal.h"
+
+namespace Director {
+namespace DT {
+
+Common::String savedStateFileName = Common::String("ImGuiSaveState.json");
+const char *WINDOW_LIST[] = {
+	"Archive", "Breakpoints", "Vars", "Watched Vars", "Score", "Control Panel",
+	"Execution Context", "Functions", "Channels", "Cast", "Settings"
+};
+
+// What are the things that need saving?
+// 1) Window Positions
+// 2) Score frame number
+// 3) Windows that are open/closed
+// 4) Watched Variables
+void saveCurrentState() {
+	Common::JSONObject json = Common::JSONObject();
+
+	// Whether windows are open or not
+	long long int openFlags = getWindowFlags();
+	json["Windows"] = new Common::JSONValue(openFlags);
+
+	// Window Settings
+	const char *windowSettings = ImGui::SaveIniSettingsToMemory();
+	json["Window Settings"] = new Common::JSONValue(windowSettings);
+
+	// Current Log
+	ImVector<char *> currentLog = _state->_logger->getItems();
+	Common::JSONArray log;
+
+	for (auto iter : currentLog) {
+		log.push_back(new Common::JSONValue(iter));
+	}
+	json["Log"] = new Common::JSONValue(log);
+
+	// Save the JSON
+	Common::JSONValue save(json);
+	debug("ImGui::Saved state: %s", save.stringify().c_str());
+
+	Common::OutSaveFile *stream = g_engine->getSaveFileManager()->openForSaving(savedStateFileName);
+
+	if (stream) {
+		stream->writeString(save.stringify());
+		stream->finalize();
+		debug("ImGui::SaveCurrentState: Saved the current ImGui State @%s", savedStateFileName.c_str());
+	} else {
+		debug("ImGui::SaveCurrentState: Failed to open the file %s for saving", savedStateFileName.c_str());
+	}
+
+	// Clean up everything
+	delete stream;
+}
+
+long long int getWindowFlags() {
+	long long int openFlags = 0;
+	openFlags += (_state->_w.archive) ? 1 : 0;
```

Comment:
I do not like these constants. That means that in order to keep compatibility, somebody must never change the order of these flags. Consider creating a small table with window names.